apex/optimizers/fused_adam.py

```diff
@@ -31,44 +53,64 @@ class FusedAdam(torch.optim.Adam):
         https://openreview.net/forum?id=ryQu7f-RZ
     """
 
-    def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
-                 weight_decay=0, amsgrad=False, eps_inside_sqrt = False):
+    def __init__(self, params,
+                 lr=1e-3, warmup=-1, t_total=-1, schedule='warmup_linear',
+                 betas=(0.9, 0.999), eps=1e-8, eps_inside_sqrt = False,
+                 weight_decay=0., max_grad_norm=0., amsgrad=False):
         if amsgrad:
             raise RuntimeError('FusedAdam does not support the AMSGrad variant.')
-        super(FusedAdam, self).__init__(params, lr, betas, eps, weight_decay, amsgrad)
+        defaults = dict(lr=lr, schedule=schedule, warmup=warmup, t_total=t_total,
+                        betas=betas, eps=eps, weight_decay=weight_decay,
+                        max_grad_norm=max_grad_norm)
+        super(FusedAdam, self).__init__(params, defaults)
         self.eps_mode = 0 if  eps_inside_sqrt else 1
 
-    def step(self, closure=None, grads=None, output_params=None, scale=1.):
+    def step(self, closure=None, grads_group=None, output_params_group=None, scale=1., grad_norms=None):
         """Performs a single optimization step.
 
         Arguments:
             closure (callable, optional): A closure that reevaluates the model
                 and returns the loss.
-            grads (list of tensors, optional): weight gradient to use for the 
-                optimizer update. If gradients have type torch.half, parameters 
+            grads (list of tensors, optional): weight gradient to use for the
+                optimizer update. If gradients have type torch.half, parameters
                 are expected to be in type torch.float. (default: None)
-            output params (list of tensors, optional): A reduced precision copy 
-                of the updated weights written out in addition to the regular 
+            output params (list of tensors, optional): A reduced precision copy
+                of the updated weights written out in addition to the regular
                 updated weights. Have to be of same type as gradients. (default: None)
             scale (float, optional): factor to divide gradient tensor values
                 by before applying to weights. (default: 1)
         """
         loss = None
         if closure is not None:
             loss = closure()
-        if grads is not None:
-           assert len(self.param_groups)==1, "mixed precision optimizer works for a single group only"
-        for group in self.param_groups:
+
+        if grads_group is None:
+            grads_group = [None]*len(self.param_groups)
+        if output_params_group is None:
+            output_params_group = [None]*len(self.param_groups)
+        if grad_norms is None:
+            grad_norms = [None]*len(self.param_groups)
+
+        for group, grads, output_params, grad_norm in zip(self.param_groups, grads_group, output_params_group, grad_norms):
```

Comment:
would this break our current usage, where we pass grads as a list for an (implicit) single group? It would need to be passed as a nested list, right?

Reply:
To make it backwards-compatible, we could check the type of grads_group[0], and wrap it in another list if it's not a list already.